File: goopang-java/src/main/java/com/example/goopangjava/service/ProductService.java
package com.example.goopangjava.service;

import com.example.goopangjava.entity.product.Clothing;
import com.example.goopangjava.entity.product.Electronics;
import com.example.goopangjava.entity.product.Product;
import com.example.goopangjava.input.AddProductInput;
import com.example.goopangjava.repository.Database;
import org.apache.coyote.BadRequestException;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Sinks;
import reactor.util.concurrent.Queues;

import java.util.List;
import java.util.UUID;

@Service
public class ProductService {

    private final Sinks.Many<Product> productSink = Sinks.many()
            .multicast()
            .onBackpressureBuffer(Queues.SMALL_BUFFER_SIZE, false);

    public Product getProduct(String productId) throws BadRequestException {
        return Database.getInstance().products.stream()
                .filter(product -> product.getId().equals(productId))
                .findFirst()
                .orElseThrow(() -> new BadRequestException("Product not found"));
    }

    public List<Product> getProducts() {
        return Database.getInstance().products;
    }

    public Product addProduct(AddProductInput addProductInput) throws BadRequestException {
        Product product = switch (addProductInput.getProductType()) {
            case ELECTRONICS -> {
                if (addProductInput.getWarrantyPeriod() == null) {
                    throw new BadRequestException("Warranty period is required");
                }
                yield new Electronics(
                        UUID.randomUUID().toString().substring(0, 5),
                        addProductInput.getName(),
                        addProductInput.getPrice(),
                        addProductInput.getWarrantyPeriod()
                );
            }
            case CLOTHING -> {
                if (addProductInput.getSize() == null) {
                    throw new BadRequestException("Size is required");
                }
                yield new Clothing(
                        UUID.randomUUID().toString().substring(0, 5),
                        addProductInput.getName(),
                        addProductInput.getPrice(),
                        addProductInput.getSize()
                );
            }
        };
        Database.getInstance().products.add(product);
        productSink.tryEmitNext(product);
        return product;
    }

    public Flux<Product> messageFlux(String productName) {
        return productSink.asFlux().filter(product -> 
            productName == null || product.getName().contains(productName)
        );
    }
}

File: goopang-kotlin/src/main/kotlin/com/example/goopang/service/ProductService.kt
package com.example.goopang.service

import com.example.goopang.entity.product.Clothing
import com.example.goopang.entity.product.Electronics
import com.example.goopang.entity.product.Product
import com.example.goopang.entity.product.ProductType
import com.example.goopang.input.AddProductInput
import com.example.goopang.repository.Database
import org.apache.coyote.BadRequestException
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Sinks
import reactor.util.concurrent.Queues
import java.util.*

@Service
class ProductService {

    private val productSink =
        Sinks
            .many()
            .multicast()
            .onBackpressureBuffer<Product>(Queues.SMALL_BUFFER_SIZE, false)

    fun getProduct(productId: String): Product {
        return Database.products.firstOrNull { it.id == productId }
            ?: throw BadRequestException("Product not found")
    }

    fun getProducts(): ArrayList<Product> {
        return Database.products
    }

    fun addProduct(addProductInput: AddProductInput): Product {
        val product = when (addProductInput.productType) {
            ProductType.ELECTRONICS -> {
                if (addProductInput.warrantyPeriod == null) {
                    throw BadRequestException("Warranty period is required")
                }
                Electronics(
                    id = UUID.randomUUID().toString().substring(0, 5),
                    name = addProductInput.name,
                    price = addProductInput.price,
                    warrantyPeriod = addProductInput.warrantyPeriod
                )
            }

            ProductType.CLOTHING -> {
                if (addProductInput.size == null) {
                    throw BadRequestException("Size id required")
                }
                Clothing(
                    id = UUID.randomUUID().toString().substring(0, 5),
                    name = addProductInput.name,
                    price = addProductInput.price,
                    size = addProductInput.size
                )
            }
        }
        Database.products.add(product)

        productSink.tryEmitNext(product)

        return product
    }

    fun messageFlux(productName: String?): Flux<Product> {
        return productSink.asFlux()
            .let {
                if (productName != null) {
                    it.filter { it.name.contains(productName) }
                } else {
                    it
                }
            }
    }

}




